src/crashreporter/crashuploader.cpp

```diff
@@ -0,0 +1,145 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "crashuploader.h"
+#include "constants.h"
+#include "qmlengineholder.h"
+#include <QApplication>
+#include <QHttpMultiPart>
+#include <QFile>
+#include <QFileInfo>
+#include <QByteArray>
+#include <QNetworkReply>
+#include "logger.h"
+
+using namespace std;
+
+constexpr auto MAX_RETRIES = 3;
+constexpr auto BOUNDARY = "--------------------XYsowmfWDDGdos";
+
+namespace {
+Logger logger(LOG_CRASHREPORTER, "CrashUploader");
+}
+
+CrashUploader::CrashUploader(QObject* parent) : QObject(parent) {
+  m_network = QmlEngineHolder::instance()->networkAccessManager();
+
+  connect(m_network, &QNetworkAccessManager::finished, this,
+          &CrashUploader::requestComplete);
+}
+
+void CrashUploader::startUploads(QStringList files) {
+  m_files = files;
+  nextUpload();
+}
+
+void CrashUploader::nextUpload() {
+  if (!m_files.empty()) {
+    m_currentFile = m_files.first();
+    m_files.pop_front();
+    m_retries = 0;
+    startRequest(m_currentFile);
+  } else {
+    emit uploadsComplete();
+  }
+}
+
+void CrashUploader::startRequest(const QString& file) {
+  logger.debug() << "Starting upload of " << file;
+  QFile dump(file);
+  if (!dump.open(QIODevice::ReadOnly)) {
+    // fail and try the next one
+    logger.error() << "Unable to open file: " << file;
+    nextUpload();
+    return;
+  }
+  m_retries++;
+  if (m_retries > MAX_RETRIES) {
+    nextUpload();
+    return;
+  }
+  auto fileName = QFileInfo(file).fileName();
+  auto dumpContent = dump.readAll();
+
+  auto multipart = new QHttpMultiPart(this);
+  multipart->setBoundary(QString(BOUNDARY).toLocal8Bit());
+  QHttpPart formPart;
+  auto contentDisp = QString(
+                         "form-data; name=\"upload_file_minidump\"; "
+                         "filename=\"%1\"")
+                         .arg(fileName);
+  formPart.setHeader(QNetworkRequest::ContentDispositionHeader,
+                     QVariant(contentDisp));
+  formPart.setHeader(QNetworkRequest::ContentLengthHeader,
+                     QVariant(dumpContent.size()));
+  formPart.setBody(dumpContent);
+  QHttpPart namePart;
+  namePart.setHeader(QNetworkRequest::ContentDispositionHeader,
+                     "form-data; name=\"ProductName\"");
+  namePart.setBody(QString("MozillaVPN").toLocal8Bit());
+
+  QHttpPart versionPart;
+  versionPart.setHeader(QNetworkRequest::ContentDispositionHeader,
+                        "form-data; name=\"Version\"");
+  versionPart.setBody(APP_VERSION);
+  multipart->append(namePart);
+  multipart->append(formPart);
+  multipart->append(versionPart);
+
+  auto urlStr = Constants::inProduction() ? Constants::CRASH_STAGING_URL
```

Comment:
I'm a little uncertain about the use of `Constants::inProduction()` here, for two reasons:
1. It's not clear to me that the crash reporter tool will be able to load the settings necessary to determine if the user was connected to staging or not (and it will just default to production).
2. The builds aren't different between staging and production (it's just a setting). So I don't think there's any value in uploading the crash reports to different locations.

Reply:
I made this change at Baku's request.  The way it should work is that is the user has set staging in the client, then we use staging for the dump, unless it is a debug build which should always use staging.  The idea is that dev mode crashes don't pollute the crash reports, and we also have some means of testing against a non-production crash collector.  If inProduction doesn't do this, let me know.  All settings should be the same as the client, as this is the same exe.

Reply:
Okay, then I'll defer to baku's judgment on this one.